From 13df8146aae61a201c6a5cfc1090092077d9336b Mon Sep 17 00:00:00 2001
From: wilsonshih <wilsonshih@google.com>
Date: Fri, 25 Oct 2019 18:47:50 +0800
Subject: [PATCH 029/297] Ensure next home activity is ready before finish
 FallbackHome

NexusLaunch could be created but haven't idle before we destroy
FallbackHome, add more state check for it.

Bug: 143199498
Test: atest ActivityRecordTests
Change-Id: Icbe436afd66d2c2e5a9b884433e2fea84c119513
(cherry picked from commit 5901007d16312cb4f183afbc8b1a35b753dcb28f)
---
 .../core/java/com/android/server/wm/ActivityRecord.java  | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/services/core/java/com/android/server/wm/ActivityRecord.java b/services/core/java/com/android/server/wm/ActivityRecord.java
index 3853841b63f..4692fcf3867 100644
--- a/services/core/java/com/android/server/wm/ActivityRecord.java
+++ b/services/core/java/com/android/server/wm/ActivityRecord.java
@@ -299,10 +299,10 @@ import android.view.WindowManager;
 import android.view.WindowManager.LayoutParams;
 import android.view.animation.Animation;
 
+import com.android.internal.R;
 import com.android.internal.annotations.VisibleForTesting;
 import com.android.internal.app.ResolverActivity;
 import com.android.internal.content.ReferrerIntent;
-import com.android.internal.R;
 import com.android.internal.util.ToBooleanFunction;
 import com.android.internal.util.XmlUtils;
 import com.android.server.AttributeCache;
@@ -2724,9 +2724,10 @@ final class ActivityRecord extends WindowToken implements WindowManagerService.A
         // TODO(b/137329632): Exclude current activity when looking for the next one with
         //  ActivityDisplay#topRunningActivity().
         final ActivityRecord next = display.topRunningActivity();
-        final boolean isLastStackOverEmptyHome =
-                next == null && stack.isFocusedStackOnDisplay() && display.getHomeStack() != null;
-        if (isLastStackOverEmptyHome) {
+        final boolean isLastStackOverUnreadyHome = display.getHomeStack() != null
+                && (next == null || (next.isActivityTypeHome() && !next.isState(RESUMED)))
+                && stack.isFocusedStackOnDisplay();
+        if (isLastStackOverUnreadyHome) {
             // Don't destroy activity immediately if this is the last activity on the display and
             // the display contains home stack. Although there is no next activity at the moment,
             // another home activity should be started later. Keep this activity alive until next
-- 
2.35.1

